core/src/main/java/io/grpc/internal/CallCredentialsApplyingTransportFactory.java

```diff
@@ -126,5 +143,50 @@ public Attributes getTransportAttrs() {
         return delegate.newStream(method, headers, callOptions);
       }
     }
+
+    @Override
+    public void shutdown(Status status) {
+      checkNotNull(status, "status");
+      synchronized (this) {
+        if (pendingApplier.get() != 0) {
+          savedShutdownStatus = status;
```

Comment:
From this point forward `newStream()` calls should produce failing streams. At it is, new streams are continued to be allowed, which means `pendingApplier` may never become `0`.

Reply:
Hm, isn't allowing new streams (accepted before ManagedChannelImpl shutdown, delayed in metadataapplier) the purpose of the fix? Otherwise if the desired behavior is  "From this point forward `realTransport.newStream()` calls should produce failing streams" , then it falls back to delay `terminating` here instead of delay `shutdown`?

Reply:
We want the `delegate` to allow new streams until this transport has drained any queued streams. But if this transport is shutdown it should not longer accept new streams; it can just continue processing the streams already created and wait for their CallCredentials to complete (at which point those streams will be created on the `delagate` and we can then shut down the delegate).

Reply:
I don't think I understand. Which `newStream()` did you mean? `callCredentialTransport.newStream()` won't be called anymore once this `shutdown()` is called.

Reply:
I was meaning `CallCredentialsApplyingTransport.newStream()`.

> callCredentialTransport.newStream() won't be called anymore once this shutdown() is called.

`newStream()` is called from many threads, so it is generally racy with `shutdown()` and the API allows `newStream()`s after `shutdown()`. Although it is unlikely to be used very long after `shutdown()`. For our own sanity detecting bugs, I think it is best to enforce the shutdown check.

I think I'd implement it by initializing the `AtomicInteger` to `Integer.MIN_VALUE`. Negative values mean "still running" and positive values mean "shutdown." During shutdown()/shutdownNow(), if the value is negative do `addAndGet(Integer.MIN_VALUE)`. When creating a new stream, if the result of `incrementAndGet()` is positive then it should create a `FailingClientStream` and call `applierListener.onComplete()`. The current `decrementAndGet()` behavior would remain the same.

If implementation complexity gets out of control, we can decide to ignore the shutdown check. But I think it won't be too bad, albeit non-obvious. That seems to be because we are doing something uncommon.